ttnn/cpp/ttnn/operations/data_movement/expand/device/expand_device_operation.cpp

```diff
@@ -0,0 +1,94 @@
+    const auto& input = tensor_args.input;
+
+    switch (input.get_layout()) {
+        case Layout::ROW_MAJOR: return ExpandRowMajorFactory{};
+        // case Layout::TILE: return ExpandTileFactory{};
+        default: TT_FATAL(false, "Unsupported input layout");
+    }
+}
+
+void validate(
+    const ExpandOperation::operation_attributes_t& operation_attributes,
+    const ExpandOperation::tensor_args_t& tensor_args) {
+    // We need to assert that the input and output are ROW_MAJOR. (unfortunately)
+
+    const auto& input = tensor_args.input;
+    const auto& output = tensor_args.output;
+
+    TT_FATAL(input.get_layout() == Layout::ROW_MAJOR, "Input tensor layout must be ROW_MAJOR");
+    if (output.has_value()) {
+        // TT_FATAL(output.value().get_layout() == Layout::ROW_MAJOR, "Output tensor layout must be ROW_MAJOR");
+        TT_FATAL(
+            output->get_shape().logical_shape().as_vector() == operation_attributes.output_shape,
+            "Output shape must match operation attributes");
+        TT_FATAL(input.get_layout() == output->get_layout(), "Input and output must have same layout");
+        TT_FATAL(input.get_dtype() == output->get_dtype(), "Input and output must have same dtype");
+        TT_FATAL(input.device() == output->device(), "Input and output must be on the same device");
+    }
+}
+
+void ExpandOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate(operation_attributes, tensor_args);
+};
+
+void ExpandOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate(operation_attributes, tensor_args);
+};
+
+ExpandOperation::shape_return_value_t ExpandOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto to_return = Shape{operation_attributes.output_shape};
+
+    // if (tensor_args.input.get_layout() == Layout::TILE) {
+    //     return to_return.padded_shape();
+    // }
+
+    return to_return.logical_shape();
+};
+
+ExpandOperation::tensor_return_value_t ExpandOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    // Let's just require it to be allocated ahead of time for now
+    if (tensor_args.output.has_value())
+        return {tensor_args.output.value()};
+
+    return create_device_tensor(
+        compute_output_shapes(operation_attributes, tensor_args),
+        tensor_args.input.get_dtype(),
+        tensor_args.input.get_layout(),
+        tensor_args.input.device(),
+        operation_attributes.output_mem_config);
+}
+
+std::tuple<ExpandOperation::operation_attributes_t, ExpandOperation::tensor_args_t> ExpandOperation::invoke(
+    const Tensor& input,
+    const std::vector<uint32_t>& output_shape,
+
+    const std::optional<Tensor>& output,
```

Comment:
```suggestion
    const std::vector<uint32_t>& output_shape,
    const std::optional<Tensor>& output,
```
I thought clang-format was executed during commit, but is that not actually the case?

Reply:
It was not. it was disabled it in a commit https://github.com/tenstorrent/tt-metal/commit/31282ff8746c81dcde3f5a8dda704f8b42218c3b

Reply:
Also this is not a change that is supposed to be done by clang-format. I added a blank line to distinguish optional kwargs and normal args. I'll remove it